### ISSUE TITLE
Potential bug: Forecasting with model.predict does not return predictions of the future

### ISSUE THREAD

**Issue by holzwolf** [alan-turing-institute/deepsensor#130]
### Description

After initialising the TaskLoader (NUMBER_OF_CONTEXT_SETS = 24, TARGET_DELTAS = 3) 
```
context_timedeltas = np.linspace(-NUMBER_OF_CONTEXT_SETS+1, 0, NUMBER_OF_CONTEXT_SETS)
target_timedeltas = np.linspace(1, TARGET_DELTAS, TARGET_DELTAS)
task_loader = TaskLoader(
    context=[normalised_df["tl"],] * NUMBER_OF_CONTEXT_SETS,
    context_delta_t=context_timedeltas,
    target=[normalised_df["tl"],] * TARGET_DELTAS,
    target_delta_t=target_timedeltas,
    time_freq="h",  # hourly frequency
) 
# context_timedeltas :  [-23. -22. -21. -20. -19. -18. -17. -16. -15. -14. -13. -12. -11. -10. -9.  -8.  -7.  -6.  -5.  -4.  -3.  -2.  -1.   0.]
# target_timedeltas : [1. 2. 3.]
```
that uses 24 h of data to forecast 1, 2 and 3 hours into the future. The training tasks have the form:

```
time: Timestamp/2014-01-02 00:00:00
 ops: []
 X_c: ['ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 242)', 'ndarray/float32/(2, 242)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)']
 Y_c: ['ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 242)', 'ndarray/float32/(1, 242)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)']
 X_t: ['ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)', 'ndarray/float32/(2, 243)']
 Y_t: ['ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)', 'ndarray/float32/(1, 243)']
```
Length of X_c and Y_c = 24,  Length of X_t and Y_t = 3.
After training a model I generate test tasks that have the form:
```
time: Timestamp/2019-01-02 10:00:00
ops: []
X_c: ['ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)', 'ndarray/float32/(2, 264)']
Y_c: ['ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)', 'ndarray/float32/(1, 264)']
X_t: None
Y_t: None
```
The X_t and Y_t are None, because I want the trained model to predict this. Therefore I use
```
model.predict(test_tasks, location)
```
And I get
```
{'tl':                                                mean       std
 time                lat       lon                            
 2019-01-02 10:00:00 47.610554 13.758333  272.719849  0.815528}
```
The issue is that I was expecting three temperature (tl) values for times 2019-01-02 11:00:00, 2019-01-02 12:00:00 and 2019-01-02 13:00:00.

Can somebody help me please how I can forecast values properly? 
Best regards

### Reproduction steps

```bash
1. Create Taskloader with a list of target_delta_t for Forecasting
2. Train a model
3. Try to use model.predict to get predictions for a single location
```

### Version

Name: deepsensor
Version: 0.3.8

### OS

Windows

**Comment by tom-andersson** [comment]
Hi @holzwolf, thanks for raising this and sorry for the delay. There are two things going on here, the first is a known annoyance with DeepSensor forecasting, and the second is a bug:
1. The `time` dimension returned by `model.predict` refers to the *initialisation time*, not the *valid time*, which is why the time dim is not 'of the future'. See: https://github.com/alan-turing-institute/deepsensor/issues/30
2. `model.predict` overwrites previous lead times of a given target variable ID, so you only get one entry regardless of whether you specify multiple lead times for a particular target variable. Forecasting has not been heavily battle-tested and this bug was an oversight on my part.

I'm working on a fix for both of the above issues now and will keep you posted.

**Comment by tom-andersson** [comment]
This should now be fixed in `deepsensor 0.4.1 `. See [this Google Colab demonstrator](https://colab.research.google.com/drive/1q1n7RGxEQJ_6z1DpBSg9FKF9bVQF-GRV#scrollTo=HkpQHoUj6JMI).

You won't need to retrain any models; just `pip install --upgrade deepsensor` and rerun `model.predict`. Note that the PR adds a `deepsensor.eval.metrics.compute_errors` method for conveniently obtaining an xarray.Dataset of pointwise forecast errors at each lead time (showcased in the demonstrator).

**Comment by davidwilby** [comment]
@all-contributors please add @holzwolf for bug
